homeassistant/auth/providers/openid.py

```diff
@@ -0,0 +1,282 @@
+        vol.Required("authorization_endpoint"): str,
+        vol.Required("response_types_supported"): vol.Contains("code"),
+        vol.Optional(
+            "token_endpoint_auth_methods_supported", default=["client_secret_basic"]
+        ): vol.Contains("client_secret_post"),
+        vol.Optional(
+            "grant_types_supported", default=["authorization_code", "implicit"]
+        ): vol.Contains("authorization_code"),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+class InvalidAuthError(HomeAssistantError):
+    """Raised when submitting invalid authentication."""
+
+
+async def raise_for_status(response: ClientResponse) -> None:
+    """Raise exception on data failure with logging."""
+    if response.status >= 400:
+        standard = ClientResponseError(
+            response.request_info,
+            response.history,
+            code=response.status,
+            headers=response.headers,
+        )
+        data = await response.text()
+        _LOGGER.error("Request failed: %s", data)
+        raise InvalidAuthError(data) from standard
+
+
+WANTED_SCOPES = {"openid", "email", "profile"}
+
+
+@AUTH_PROVIDERS.register("openid")
+class OpenIdAuthProvider(AuthProvider):
+    """Example auth provider based on hardcoded usernames and passwords."""
+
+    DEFAULT_TITLE = "OpenID Connect"
+
+    _configuration: Dict[str, Any]
+    _jwks: Dict[str, Any]
+
+    async def async_get_configuration(self) -> Dict[str, Any]:
+        """Get discovery document for OpenID."""
+        session = async_get_clientsession(self.hass)
+        async with session.get(self.config[CONF_CONFIGURATION]) as response:
+            await raise_for_status(response)
+            data = await response.json()
+        return cast(Dict[str, Any], OPENID_CONFIGURATION_SCHEMA(data))
+
+    async def async_get_jwks(self) -> Dict[str, Any]:
+        """Get the keys for id verification."""
+        session = async_get_clientsession(self.hass)
+        async with session.get(self._configuration["jwks_uri"]) as response:
+            await raise_for_status(response)
+            data = await response.json()
+        return cast(Dict[str, Any], data)
+
+    @property
+    def redirect_uri(self) -> str:
+        """Return the redirect uri."""
+        url = URL(get_url(self.hass, prefer_external=True, allow_cloud=False))
+        return str(url.with_path(AUTH_CALLBACK_PATH))
+
+    async def async_login_flow(self, context: Optional[Dict]) -> LoginFlow:
+        """Return a flow to login."""
+
+        if not hasattr(self, "_configuration"):
+            self._configuration = await self.async_get_configuration()
+
+        if not hasattr(self, "_jwks"):
+            self._jwks = await self.async_get_jwks()
+
+        async_register_view(self.hass)
+
+        return OpenIdLoginFlow(self)
+
+    async def async_retrieve_token(self, code: str) -> Dict[str, Any]:
+        """Convert a token code into an actual token."""
+
+        payload = {
```

Comment:
We should use LocalOAUth2Implementaiton for this https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/config_entry_oauth2_flow.py#L88

Reply:
I have WIP on that. It does mean more changes to those classes. Somewhat like this: https://github.com/elupus/home-assistant/compare/openid...elupus:openid_standard_api

Reply:
Also can be added. I'm not sure there was much benefit to be honest. (well unless we can hook this up to cloud and use it's stored client_id/secret)

Reply:
Anyway, i added those changes to this branch.